lib/core/test/unit/Cardano/Wallet/DB/PoolProperties.hs

```diff
@@ -0,0 +1,111 @@
+    ( poolProperties
+    , withPoolDB
+    , newMemoryPoolDBLayer
+    ) where
+
+import Prelude
+
+import Cardano.BM.Configuration.Model
+    ( Configuration )
+import Cardano.BM.Data.LogItem
+    ( LogObject (..) )
+import Cardano.BM.Trace
+    ( traceInTVarIO )
+import Cardano.DB.Sqlite
+    ( SqliteContext )
+import Cardano.Pool.DB
+    ( DBLayer (..) )
+import Cardano.Pool.DB.Sqlite
+    ( newDBLayer )
+import Cardano.Wallet.DB.Arbitrary
+    ( StakePoolsFixture (..), once )
+import Cardano.Wallet.Primitive.Types
+    ( PoolId, SlotId (..) )
+import Cardano.Wallet.Unsafe
+    ( unsafeRunExceptT )
+import Control.Monad
+    ( forM_ )
+import Control.Monad.IO.Class
+    ( liftIO )
+import Data.Functor
+    ( ($>) )
+import Data.List.Extra
+    ( nubOrd )
+import Data.Text
+    ( Text )
+import GHC.Conc
+    ( TVar, newTVarIO )
+import Test.Hspec
+    ( Spec, SpecWith, beforeAll, beforeWith, describe, it, shouldReturn )
+import Test.QuickCheck
+    ( Gen, Property, elements, property )
+import Test.QuickCheck.Monadic
+    ( monadicIO )
+
+import qualified Cardano.Pool.DB.MVar as MVar
+import qualified Data.Map.Strict as Map
+
+-- | Provide a DBLayer to a Spec that requires it. The database is initialised
+-- once, and cleared with 'cleanDB' before each test.
+withPoolDB :: IO (DBLayer IO) -> SpecWith (DBLayer IO) -> Spec
+withPoolDB create = beforeAll create . beforeWith (\db -> cleanDB db $> db)
+
+-- | Set up a DBLayer for testing, with the command context, and the logging
+-- variable.
+newMemoryPoolDBLayer :: IO Configuration -> IO (DBLayer IO)
+newMemoryPoolDBLayer conf = snd . snd <$> (newMemoryPoolDBLayer' conf)
+
+newMemoryPoolDBLayer'
+    :: IO Configuration -> IO (TVar [LogObject Text], (SqliteContext, DBLayer IO))
+newMemoryPoolDBLayer' testingLogConfig = do
+    logConfig <- testingLogConfig
+    logVar <- newTVarIO []
+    (logVar, ) <$> newDBLayer logConfig (traceInTVarIO logVar) Nothing
+
+poolProperties :: SpecWith (DBLayer IO)
+poolProperties = do
+    describe "Stake Pool properties" $ do
+        it "putPoolProduction . readPoolProduction yields expected results"
+            (property . prop_putReadPoolProduction)
+
+{-------------------------------------------------------------------------------
+                                    Properties
+-------------------------------------------------------------------------------}
+
+-- | Can read put pool production
+prop_putReadPoolProduction
+    :: DBLayer IO
+    -> StakePoolsFixture
+    -> Property
+prop_putReadPoolProduction db (StakePoolsFixture pairs) =
+    monadicIO (setup >>= prop)
+  where
+    epochs = nubOrd $ map (epochNumber . snd) pairs
+    setup = do
+        liftIO (cleanDB db)
+        liftIO MVar.newDBLayer
+    prop db' = liftIO $ do
+        _ <- once pairs $ \(pool, slot) ->
+            unsafeRunExceptT $ putPoolProduction db slot pool
+        _ <- once pairs $ \(pool, slot) ->
+            unsafeRunExceptT $ putPoolProduction db' slot pool
```

Comment:
Wouldn't it be easier (and more correct) to only generate unique pairs :thinking: ? `once` was a useful combinator in properties where we generated conflicting entries **on purpose** because they were the point of the test. Here, I think the pairs should really be generated unique (and I think they already are).

Reply:
indeed, we generate pairs uniquely, so `forM_` is ok here